pkg/nacler/keyers/secureenclave/secureenclave.m

```diff
@@ -0,0 +1,239 @@
+#import <Foundation/Foundation.h>
+#import <CoreFoundation/CoreFoundation.h>
+#import <Security/Security.h>
+#import "secureenclave.h"
+
+
+// CFStringToCString creates a copy of the bytes of data
+// converted to UTF-8 encoding.
+char* CFStringToCString(CFStringRef data) {
+  CFIndex len = CFStringGetLength(data);
+  if (!len)
+    return NULL;
+
+  CFIndex realLen = CFStringGetMaximumSizeForEncoding(len, kCFStringEncodingUTF8);
+  char* buf = (char*)malloc((size_t)realLen + 1);
+  if (!buf)
+    return NULL;
+  memset(buf, 0, (size_t)realLen + 1);
+
+  Boolean ok = CFStringGetCString(data, buf, realLen + 1, kCFStringEncodingUTF8);
+  if (!ok)
+    return NULL;
+
+  return buf;
+}
+
+unsigned char* CFDataToUint8(CFDataRef data) {
+  CFIndex len = CFDataGetLength(data);
+  if (!len)
+    return NULL;
+
+  UInt8* buf = (UInt8*)malloc((size_t)len);
+  if (!buf)
+    return NULL;
+  memset(buf, 0, (size_t)len);
+
+  CFRange range = CFRangeMake(0, len);
+  CFDataGetBytes(data, range, buf);
+
+  return (unsigned char*)buf;
+}
+
+// Extract the public key data from a SecKeyRef
+// Returns null if it couldn't find the data
+CFDataRef ExtractPubKey(SecKeyRef pubKey) {
+  CFDataRef val = NULL;
+  CFDataRef res = NULL;
+  CFDictionaryRef keyAttrs = SecKeyCopyAttributes(pubKey);
+  if (CFDictionaryContainsKey(keyAttrs, kSecValueData) == true)
+    val = (CFDataRef)CFDictionaryGetValue(keyAttrs, kSecValueData);
+
+  if (val)
+    res = CFDataCreateCopy(kCFAllocatorDefault, val);
+
+  if (keyAttrs)
+    CFRelease((CFTypeRef)keyAttrs);
+
+  return res;
+}
+
+size_t createKey(unsigned char** ret, char** retErr){
+    CFErrorRef error = NULL;
+
+    // ignoring the depreciated warning becuase this is the setting we need to be able to access secure enclave
+    // for this particualar key even when device is locked, this was depreciated without providing an alternative
```

Comment:
Maybe add a note that the deprecated setting is `kSecAttrAccessibleAlwaysThisDeviceOnly`.

What will happen if a user does migrate to a new device? Would we require/expect a full reinstall of launcher?

Reply:
called out that setting in the comment

> What will happen if a user does migrate to a new device?

It's a new device, they'll be forced to register in K2, this is the goal

> Would we require/expect a full reinstall of launcher?

Yes, I think so.